File: back/DevC/src/main/java/com/back/devc/domain/admin/dashboard/service/AdminDashboardService.java
package com.back.devc.domain.admin.dashboard.service;

import com.back.devc.domain.admin.dashboard.dto.DashboardResponseDto;
import com.back.devc.domain.admin.dashboard.dto.DashboardResponseDto.ReportReasonCount;
import com.back.devc.domain.admin.dashboard.repository.AdminDashboardRepository;
import com.back.devc.domain.interaction.report.entity.ReportStatus;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;

@Service
@RequiredArgsConstructor
public class AdminDashboardService {

    private final AdminDashboardRepository adminDashboardRepository;

    public DashboardResponseDto getDashboardData() {

        LocalDate today = LocalDate.now();

        // =========================
        // 1. 전체 통계
        // =========================
        Long totalUsers = adminDashboardRepository.countAllUsers();
        Long totalPosts = adminDashboardRepository.countAllPosts();
        Long pendingReports = adminDashboardRepository.countPendingReports();

        // =========================
        // 2. 오늘 활동
        // =========================
        Long newUsers = adminDashboardRepository.countTodayUsers(today);
        Long newPosts = adminDashboardRepository.countTodayPosts(today);
        Long newComments = adminDashboardRepository.countTodayComments(today);

        // =========================
        // 3. 게시글 신고
        // =========================
        Long postReportTotal = adminDashboardRepository.countTodayPostReports(today);
        Long postReportPending = adminDashboardRepository.countTodayPostReportsByStatus(today, ReportStatus.PENDING);
        Long postReportResolved = adminDashboardRepository.countTodayPostReportsByStatus(today, ReportStatus.RESOLVED);
        java.util.List<ReportReasonCount> postReportReasons =
                adminDashboardRepository.countTodayPostReportsByReason(today);

        // =========================
        // 4. 댓글 신고
        // =========================
        Long commentReportTotal = adminDashboardRepository.countTodayCommentReports(today);
        Long commentReportPending = adminDashboardRepository.countTodayCommentReportsByStatus(today, ReportStatus.PENDING);
        Long commentReportResolved = adminDashboardRepository.countTodayCommentReportsByStatus(today, ReportStatus.RESOLVED);
        java.util.List<ReportReasonCount> commentReportReasons =
                adminDashboardRepository.countTodayCommentReportsByReason(today);

        // =========================
        // 5. Response 조립
        // =========================
        return DashboardResponseDto.builder()
                .summary(DashboardResponseDto.SummaryStats.builder()
                        .totalUsers(totalUsers)
                        .totalPosts(totalPosts)
                        .pendingReports(pendingReports)
                        .todayVisitors(0L) // 아직 미구현이면 고정
                        .build())
                .todayActivity(DashboardResponseDto.TodayActivity.builder()
                        .newUsers(newUsers)
                        .newPosts(newPosts)
                        .newComments(newComments)
                        .build())
                .todayReports(DashboardResponseDto.TodayReportStats.builder()
                        .post(DashboardResponseDto.ReportCategory.builder()
                                .total(postReportTotal)
                                .pending(postReportPending)
                                .resolved(postReportResolved)
                                .byReason(postReportReasons)
                                .build())
                        .comment(DashboardResponseDto.ReportCategory.builder()
                                .total(commentReportTotal)
                                .pending(commentReportPending)
                                .resolved(commentReportResolved)
                                .byReason(commentReportReasons)
                                .build())
                        .build())
                .build();
    }
}

File: back/DevC/src/main/kotlin/com/back/devc/domain/admin/dashboard/service/AdminDashboardService.kt
package com.back.devc.domain.admin.dashboard.service

import com.back.devc.domain.admin.dashboard.dto.DashboardResponseDto
import com.back.devc.domain.admin.dashboard.dto.DashboardResponseDto.ReportCategory
import com.back.devc.domain.admin.dashboard.dto.DashboardResponseDto.ReportReasonCount
import com.back.devc.domain.admin.dashboard.repository.AdminDashboardRepository
import com.back.devc.domain.interaction.report.entity.ReportStatus
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Clock
import java.time.LocalDate

@Service
@Transactional(readOnly = true)
class AdminDashboardService {
    private val adminDashboardRepository: AdminDashboardRepository
    private val clock: Clock

    @Autowired
    constructor(adminDashboardRepository: AdminDashboardRepository) : this(
        adminDashboardRepository,
        Clock.systemDefaultZone(),
    )

    constructor(
        adminDashboardRepository: AdminDashboardRepository,
        clock: Clock,
    ) {
        this.adminDashboardRepository = adminDashboardRepository
        this.clock = clock
    }

    fun getDashboardData(): DashboardResponseDto {
        val today = LocalDate.now(clock)
        val postReportReasons = adminDashboardRepository.countTodayPostReportsByReason(today)
        val commentReportReasons = adminDashboardRepository.countTodayCommentReportsByReason(today)

        return DashboardResponseDto(
            summary = DashboardResponseDto.SummaryStats(
                totalUsers = adminDashboardRepository.countAllUsers(),
                totalPosts = adminDashboardRepository.countAllPosts(),
                pendingReports = adminDashboardRepository.countPendingReports(),
                todayVisitors = 0L,
            ),
            todayActivity = DashboardResponseDto.TodayActivity(
                newUsers = adminDashboardRepository.countTodayUsers(today),
                newPosts = adminDashboardRepository.countTodayPosts(today),
                newComments = adminDashboardRepository.countTodayComments(today),
            ),
            todayReports = DashboardResponseDto.TodayReportStats(
                post = buildPostReportCategory(today, postReportReasons),
                comment = buildCommentReportCategory(today, commentReportReasons),
            ),
        )
    }

    private fun buildPostReportCategory(
        today: LocalDate,
        byReason: List<ReportReasonCount>,
    ): ReportCategory {
        return ReportCategory(
            total = adminDashboardRepository.countTodayPostReports(today),
            pending = adminDashboardRepository.countTodayPostReportsByStatus(today, ReportStatus.PENDING),
            resolved = adminDashboardRepository.countTodayPostReportsByStatus(today, ReportStatus.RESOLVED),
            byReason = byReason,
        )
    }

    private fun buildCommentReportCategory(
        today: LocalDate,
        byReason: List<ReportReasonCount>,
    ): ReportCategory {
        return ReportCategory(
            total = adminDashboardRepository.countTodayCommentReports(today),
            pending = adminDashboardRepository.countTodayCommentReportsByStatus(today, ReportStatus.PENDING),
            resolved = adminDashboardRepository.countTodayCommentReportsByStatus(today, ReportStatus.RESOLVED),
            byReason = byReason,
        )
    }
}
